### PR TITLE
feat: implement StarkNet Pedersen hash and ECDSA

### LINKED ISSUE #15: Implement Pedersen hash
We need to implement the equivalent of [pedersen_hash](https://github.com/starkware-libs/cairo-lang/blob/7712b21fc3b1cb02321a58d0c0579f5370147a8b/src/starkware/crypto/starkware/crypto/signature/fast_pedersen_hash.py#L34) from [cairo-lang](https://github.com/starkware-libs/cairo-lang) as we'll need it for the signer.

### PR DESCRIPTION
**Pull request by xJonathanLEI**
Resolves #15.

Supersedes #18.

This PR implements Pedersen hash and Stark ECDSA with a new crate `starknet-crypto`. The implementation is inspired by the following projects:

- [eqlabs/pathfinder](https://github.com/eqlabs/pathfinder)
- [starkware-libs/crypto-cpp](https://github.com/starkware-libs/crypto-cpp)
- [starkware-libs/cairo-lang](https://github.com/starkware-libs/cairo-lang)

Since the new crate implements low-level crypto functions, to prevent accidental misuse, the module is _NOT_ re-exported. Instead, a future PR will implement high-level wrappers in `starknet-core` that use these low-level primitives under the hood.

### CODE REVIEW

**Review by milancermak — APPROVED** [review]
Awesome stuff this, good job!

**Comment by milancermak** [comment]
Heh, so eqlabs just did some updates to the curve calc in pathfinder. Do you want to incorporate them?

File: starknet-crypto/src/pedersen_hash.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+///
+/// * `x`: The x coordinate
+/// * `y`: The y coordinate
+pub fn pedersen_hash(x: &FieldElement, y: &FieldElement) -> FieldElement {
+    let mut result = SHIFT_POINT;
+    let x = x.into_bits();
+    let y = y.into_bits();
+
+    // Add a_low * P1
+    let tmp = PEDERSEN_P0.multiply(&x[..248]);
```
[the remark this answers is not in the document]
Comment (xJonathanLEI) [reply starknet-crypto/src/pedersen_hash.rs:21]: Ah I forgot to update the comments.

The change of naming was actually done to be aligned with `cairo-lang`.

**Comment by xJonathanLEI** [comment]
Yep they just mentioned it in the Telegram group. Lemme go ahead and port over to see the impact on performance. Very excited!

**Pull request merged**